### PR TITLE
Added basic architecture for common tests

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

## Summary

This PR is a simplifycation of this other PR https://github.com/gazebosim/gz-physics/pull/354

This draft PR try to set up a common arquitecture to test all physics engine under the same tests avoiding duplication and improving coverage.

## Test it

```bash
colcon test --merge-install --event-handlers console_direct+ --packages-select ignition-physics6 --ctest-args -R COMMON
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by mjcarroll** [comment]
I like this, it's very much in line with what we discussed.

@scpeters what do you think?

File: test/test_common_config.h.in [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#define PROJECT_SOURCE_PATH "${PROJECT_SOURCE_DIR}"
+#define PROJECT_BUILD_PATH "${PROJECT_BINARY_DIR}"
+#define GZ_PHYSICS_TEST_PLUGIN_PATH "${CMAKE_BINARY_DIR}/lib"
+
+/// \brief Helper macro to instantiate gtest for different engines
+#define PHYSICS_ENGINE_VALUES ::testing::ValuesIn(\
+    gz::physics::TestValues())
+
+static const std::vector<const char *> kPhysicsEngineTestValues
+  {"bullet", "dartsim", "tpe"};
```
Comment (scpeters) [inline test/test_common_config.h.in:13]: this reminds me of gazebo11's approach

File: test/common_test/basic_test.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+{
+  auto engine =
+    gz::physics::RequestEngine3d<Features>::From(physicsPlugin);
+
+  ASSERT_NE(nullptr, engine);
+  EXPECT_TRUE(engine->GetName().find(GetParam()) != std::string::npos);
+}
+
+INSTANTIATE_TEST_CASE_P(EntityManagementFeatures, EntityManagementFeaturesTest,
+    PHYSICS_ENGINE_VALUES,
```
Comment (scpeters) [inline test/common_test/basic_test.cc:74]: this reminds me of gazebo11's approach

**Comment by ahcorde** [comment]
@steve what do you think about this logic ?

File: test/common_test/basic_test.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+class EntityManagementFeaturesTest:
+  public testing::Test,
+  public testing::WithParamInterface<const char *>
+{
+  // Documentation inherited
+  public: void SetUp() override
+  {
+    gz::common::Console::SetVerbosity(4);
+
+    if (!gz::common::env("PHYSICS_ENGINE_NAME", physicsEngineName))
```
Comment (scpeters) [inline test/common_test/basic_test.cc:36]: I think we shouldn't need to specify `PHYSICS_ENGINE_NAME`. If it's not specified, we should use `FindFeatures3d` to identify the plugin names contained within the `LIB_TO_TEST` and then test each of those plugins. To support this, we should get the list of pluginNames in `SetUp` and move the `gz::plugin::Loader` instantiation and `LoadLib` call to a loop in the test case

* https://github.com/gazebosim/gz-physics/blob/e91c58a49522cfdc5793fe8b980566eb26531ba4/examples/hello_world_loader/hello_world_loader.cc#L51-L63

**Review by scpeters — COMMENTED** [review]
environment variables are a little easier to set from cmake, but if possible I think it would be nice to use CLI arguments to specify the library name and optional plugin name. We can save that for a later refactoring though

File: test/common_test/basic_test.cc (hunk 3 above)
Comment (ahcorde) [reply test/common_test/basic_test.cc]: https://github.com/gazebosim/gz-physics/pull/357/commits/1f48f3aead7cfea6948fb4d980e7c3f840badadb

**Comment by ahcorde** [comment]
@scpeters I'm not really sure about adding command line arguments to gtest. I will take a look

**Comment by ahcorde** [comment]
@scpeters I added this draft PR https://github.com/gazebosim/gz-physics/pull/359 with the changes required to use cmd line arguments

File: test/common_test/basic_test.cc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+#include <gz/physics/RequestEngine.hh>
+
+// The features that an engine must have to be loaded by this loader.
+using Features = gz::physics::FeatureList<
+  gz::physics::GetEngineInfo
+>;
+
+class EntityManagementFeaturesTest:
+  public testing::Test,
+  public testing::WithParamInterface<const char *>
```
Comment (scpeters) [inline test/common_test/basic_test.cc:35]: nit: we aren't currently using parameterized tests (`TEST_P`) so this `WithParamInterface` line can be removed

File: test/common_test/basic_test.cc [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include <gtest/gtest.h>
+
+#include <gz/common/Filesystem.hh>
+#include <gz/common/Console.hh>
```
Comment (scpeters) [inline test/common_test/basic_test.cc:20]: nit: alphabetize

File: test/common_test/basic_test.cc [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    std::string libToTest;
+    if (!gz::common::env("LIB_TO_TEST", libToTest))
+    {
+      FAIL();
+    }
+
+    auto plugins = loader.LoadLib(libToTest);
+
+    pluginNames = gz::physics::FindFeatures3d<Features>::From(loader);
```
Comment (scpeters) [inline test/common_test/basic_test.cc:50]: future todo: we should also run the `3f`, `2d`, and `2f` variants of `FindFeatures`

**Review by scpeters — COMMENTED** [review]
this is great, and I also like the changes in #357

**Review by scpeters — APPROVED** [review]

**Comment by ahcorde** [comment]
@scpeters Do you mind to review this PR https://github.com/gazebosim/gz-physics/pull/359 which built on top of this one?

**Comment by scpeters** [comment]
> @scpeters Do you mind to review this PR #359 which built on top of this one?

I just approved it 😄

**Pull request merged**